k2/csrc/fsa_equivalent_test.cc

```diff
@@ -0,0 +1,61 @@
+// k2/csrc/fsa_equivalent_test.cc
+
+// Copyright (c)  2020  Haowen Qiu
+
+// See ../../LICENSE for clarification regarding multiple authors
+
+#include "k2/csrc/fsa_equivalent.h"
+
+#include <utility>
+#include <vector>
+
+#include "gtest/gtest.h"
+#include "k2/csrc/fsa.h"
+
+namespace k2 {
+
+TEST(Properties, RandomPathFail) {
+  {
+    Fsa fsa;
+    Fsa path;
+    bool status = RandomPath(fsa, &path);
+    EXPECT_FALSE(status);
+  }
+  // TODO(haowen): add tests for non-connected fsa
+}
+
+TEST(Properties, RandomPathSuccess) {
+  {
+    Fsa fsa;
+    std::vector<Arc> arcs = {
+        {0, 1, 1}, {0, 2, 2}, {1, 2, 3}, {2, 3, 4},
+        {2, 4, 5}, {3, 4, 7}, {4, 5, 9},
+    };
+    std::vector<int32_t> arc_indexes = {0, 2, 3, 5, 6, 7};
+    fsa.arc_indexes = std::move(arc_indexes);
+    fsa.arcs = std::move(arcs);
+    Fsa path;
+
+    {
+      bool status = RandomPath(fsa, &path);
+      EXPECT_TRUE(status);
+    }
+
+    {
+      for (size_t i = 0; i < 20; ++i) {
```

Comment:
```int32_t i = 0;```

or

```auto i = 0;```

Reply:
This guy, Walter E. Brown (https://cppcon2019.sched.com/speaker/walter_e_brown)
suggests using
```cpp
for (auto i = 0; i != 20; ++i)
```

**!=** instead of **<**

---

The motivation is explained in this presentation
https://www.youtube.com/watch?v=OAmWHmwlMwI

At `37:57`, i.e., 37 minutes, 57 seconds.

Reply:
change to `auto` and `!=` (but I still think `<` is better here as we don't really care about how many times the loop will run.)

Reply:
@csukuangfj  next time please summarize the reason in text, was hard to get to the video!
Sorry I don't buy this.  My problem is that it doesn't work when you increment by 2, and it doesn't work in cases where there was less than an empty range, e.g. if begin > end (in these cases you generally want to do nothing, not loop forever and segfault).